### PR TITLE
#883 Resolved issue with RelationshipErrorHandler producing malformed output

### LINKED ISSUE #883: RelationshipErrorHandlerFactory fails with "Data at the root level is invalid"
Using RelationshipErrorHandlerFactory.Rewrite with an output that is shorter than the original URI causes the output XML to be malformed in this Word document.

```
        static void Main(string[] args)
        {
            var doc = WordprocessingDocument.Open("BrokenURI.docx", isEditable: true, new OpenSettings
            {
                RelationshipErrorHandlerFactory = _ => new RemoveMalformedHyperlinksRelationshipErrorHandler(),
            });
        }

        private sealed class RemoveMalformedHyperlinksRelationshipErrorHandler : RelationshipErrorHandler
        {
            // Works
            //public override string Rewrite(Uri partUri, string id, string uri) => $"https://error{new string('r', uri.Length)}";

            // Fails with "Data at the root level is invalid"
            public override string Rewrite(Uri partUri, string id, string uri) => $"https://error";
        }
```

When the above is run on a Word document with the following document.xml.rels:

```
<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">
	<Relationship Id="rId8" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/fontTable" Target="fontTable.xml"/>
	<Relationship Id="rId3" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/settings" Target="settings.xml"/>
	<Relationship Id="rId7" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/hyperlink" Target="mailto:test@test.com;%20test2@test.com;%252test3@test.com;%20test3@test.com;%20test4@test.com;%20test5@test.com?subject=Unsubscribe%20Request&amp;body=Please%20unsubscribe%20me%20from%20all%20future%20communications" TargetMode="External"/>
	<Relationship Id="rId2" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/styles" Target="styles.xml"/>
	<Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/numbering" Target="numbering.xml"/>
	<Relationship Id="rId6" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/endnotes" Target="endnotes.xml"/>
	<Relationship Id="rId5" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/footnotes" Target="footnotes.xml"/>
	<Relationship Id="rId4" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/webSettings" Target="webSettings.xml"/>
	<Relationship Id="rId9" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/theme" Target="theme/theme1.xml"/>
</Relationships>
```
It fails with "Data at the root level is invalid. Line 12, position 17." and has produced the following in the output document:

```
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">
  <Relationship Id="rId8" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/fontTable" Target="fontTable.xml" />
  <Relationship Id="rId3" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/settings" Target="settings.xml" />
  <Relationship Id="rId7" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/hyperlink" Target="https://error" TargetMode="External" />
  <Relationship Id="rId2" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/styles" Target="styles.xml" />
  <Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/numbering" Target="numbering.xml" />
  <Relationship Id="rId6" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/endnotes" Target="endnotes.xml" />
  <Relationship Id="rId5" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/footnotes" Target="footnotes.xml" />
  <Relationship Id="rId4" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/webSettings" Target="webSettings.xml" />
  <Relationship Id="rId9" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/theme" Target="theme/theme1.xml" />
</Relationships>l"/><Relationship Id="rId9" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/theme" Target="theme/theme1.xml"/></Relationships>
```

As you can see, the end part of the XML has now been malformed. This is probably a buffer issue with the shorter string being returned from the Rewrite method. I've seen this in a few places when working with the Package classes.

Extending the Rewrite output to match the length of the original URI resolves the issue:

`public override string Rewrite(Uri partUri, string id, string uri) => $"https://error{new string('r', uri.Length)}";`

**Information**

- .NET Target: .NET Core
- DocumentFormat.OpenXml Version: 2.12.1

Please see example console app with sample Word document repo'ing the issue:

https://github.com/Muppets/RelationshipErrorHandlerFactoryTest

**Observed**

Exception thrown: System.Xml.XmlException: 'Data at the root level is invalid. Line 12, position 17.'

**Expected**

Invalid URI is replaced with https://error

Just as a footnote, I'm loving the new RelationshipErrorHandlerFactory support added recently, great work! 🎉

### PR DESCRIPTION
**Pull request by Muppets**
I played around with a bunch of different ways to try to fix this and this was the cleanest I could come up with. Setting the `stream.Position = 0` didn't help. It seems nothing sets the original stream's length to the new smaller amount of bytes. Open to better ways of fixing this.

I've updated malformed_uri.xlsx to include a super long URI and updated the tests to match to ensure we catch the issue.

Fixes #883

### CODE REVIEW

File: src/DocumentFormat.OpenXml/Packaging/RelationshipErrorHandler.cs [hunk 1]
```diff
@@ -79,7 +79,12 @@ internal void Handle(Package package)
                     {
                         using var stream = part.GetStream(FileMode.Open, FileAccess.Write);
```
Comment (twsouthwick) [inline src/DocumentFormat.OpenXml/Packaging/RelationshipErrorHandler.cs:80]: What about keeping it as it was, but opening it with `FileMode.Truncate` (see https://docs.microsoft.com/en-us/dotnet/api/system.io.filemode?view=net-5.0).

File: test/DocumentFormat.OpenXml.Tests/OpenXmlDomTest/DocumentOpenTests.cs [hunk 2]
```diff
@@ -60,7 +60,7 @@ public void RewriteMalformedUri()
                         Count++;
                         Assert.Equal(Id, id);
                         Assert.Equal("/xl/worksheets/_rels/sheet1.xml.rels", partUri.OriginalString);
-                        Assert.Equal("mailto:one@", uri);
+                        Assert.Equal("mailto:test@test.com;%20test2@test.com;%252test3@test.com;%20test3@test.com;%20test4@test.com;%20test5@test.com?subject=Unsubscribe%20Request&body=Please%20unsubscribe%20me%20from%20all%20future%20communications", uri);
```
Comment (twsouthwick) [inline test/DocumentFormat.OpenXml.Tests/OpenXmlDomTest/DocumentOpenTests.cs:63]: Can you keep the original test and add a new one? including the updated malformed_uri.xlsx in addition to the original one

**Review by twsouthwick — CHANGES_REQUESTED** [review]

**Comment by twsouthwick** [comment]
btw the test failure doesn't seem to be showing up here, but it's an ObjectDisposedException on .NET 3.5. By switching things to truncate should fix that.

File: src/DocumentFormat.OpenXml/Packaging/RelationshipErrorHandler.cs (hunk 1 above)
Comment (Muppets) [reply src/DocumentFormat.OpenXml/Packaging/RelationshipErrorHandler.cs:80]: Not supported it seems 😢 

```
    using var stream = part.GetStream(FileMode.Truncate, FileAccess.Write);
    doc.Save(stream);
```

```
  Message: 
    System.ArgumentException : Truncate not supported
  Stack Trace: 
    PackagePart.GetStream(FileMode mode, FileAccess access)
    RelationshipErrorHandler.Handle(Package package) line 80
    OpenXmlPackage.Load(Package _package) line 83
    OpenXmlPackage.ctor(PackageLoader& loader, OpenSettings settings) line 53
    SpreadsheetDocument.ctor(PackageLoader& loader, OpenSettings settings) line 70
    SpreadsheetDocument.Open(Stream stream, Boolean isEditable, OpenSettings openSettings) line 298
    DocumentOpenTests.RewriteMalformedUri() line 69
```

File: test/DocumentFormat.OpenXml.Tests/OpenXmlDomTest/DocumentOpenTests.cs (hunk 2 above)
Comment (Muppets) [reply test/DocumentFormat.OpenXml.Tests/OpenXmlDomTest/DocumentOpenTests.cs:63]: Yep yep, let me move that out.

File: src/DocumentFormat.OpenXml/Packaging/RelationshipErrorHandler.cs [hunk 3]
```diff
@@ -79,7 +79,13 @@ internal void Handle(Package package)
...
 
-                        doc.Save(stream);
+                        // Saving directly to the zip stream can cause issues if there are less bytes being written than the original
+                        using var ms = new MemoryStream();
+                        using var writer = new StreamWriter(ms);
+                        doc.Save(writer);
+                        stream.SetLength(ms.Length);
+                        ms.Position = 0;
+                        ms.CopyTo(stream);
```
Comment (twsouthwick) [inline src/DocumentFormat.OpenXml/Packaging/RelationshipErrorHandler.cs:88]:
```suggestion
                        stream.SetLength(0);
                        doc.Save(stream);
```

Comment (twsouthwick) [reply src/DocumentFormat.OpenXml/Packaging/RelationshipErrorHandler.cs:88]: I cloned this locally and got it working by just calling `.SetLength(0)`.

**Review by twsouthwick — CHANGES_REQUESTED** [review]

File: src/DocumentFormat.OpenXml/Packaging/RelationshipErrorHandler.cs (hunk 3 above)
Comment (Muppets) [reply src/DocumentFormat.OpenXml/Packaging/RelationshipErrorHandler.cs:88]: Oh why didn't I think of that?! Love it.

**Review by twsouthwick — APPROVED** [review]
LGTM!

I'll get a 2.12.2 release out that has this fix.

**Pull request merged**